### DISCUSSION TITLE
DynUNet pipeline with data outside Decathlon dataset

### DISCUSSION THREAD

**Discussion by bip5** [Project-MONAI/tutorials#769] [Q&A] [answered]
Hi guys, I'm trying to use the DynUNet pipeline for BraTS' 21 dataset. However I'm finding the process is more involved than I expected it to be as the data pipeline is designed specifically for the decathlon dataset. I've started by modifying the datalist to be in the same format as the output from the built in decathlon loader but it feels like I'm going about this the wrong way.

Edit: I've now transferred my data into the required format except for changing the labels from (0,1,2,4] to (0,1,2,3). Is this necessary with the tutorial pipeline? Which portion of the pipeline do I need to change to achieve this switch at data loading?

Currently I'm getting an index out of bounds error and expecting this to be source of issue.

> /pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:312: operator(): block: [3022,0,0], thread: [63,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
2022-07-06 22:00:20,656 - ignite.engine.engine.DynUNetTrainer - ERROR - Current run is terminating due to exception: cuDNN error: CUDNN_STATUS_NOT_INITIALIZED
2022-07-06 22:00:20,657 - ignite.engine.engine.DynUNetTrainer - ERROR - Exception: cuDNN error: CUDNN_STATUS_NOT_INITIALIZED
Traceback (most recent call last):
  File "/home/a.bip5/.conda/envs/pix2pix/lib/python3.9/site-packages/ignite/engine/engine.py", line 840, in _run_once_on_dataset
    self.state.output = self._process_function(self, self.state.batch)
  File "/scratch/a.bip5/BraTS 2021/nnUNet/./nnu.py", line 830, in _iteration
    engine.state.output[Keys.LOSS].backward()
  File "/home/a.bip5/.conda/envs/pix2pix/lib/python3.9/site-packages/torch/_tensor.py", line 255, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
  File "/home/a.bip5/.conda/envs/pix2pix/lib/python3.9/site-packages/torch/autograd/__init__.py", line 147, in backward
    Variable._execution_engine.run_backward(
RuntimeError: cuDNN error: CUDNN_STATUS_NOT_INITIALIZED

Thank you!

**Comment by Nic-Ma** [comment]
Hi @yiheng-wang-nv ,

Could you please help provide some comments for this question?

Thanks in advance.

**Comment by yiheng-wang-nv** [comment, marked as answer]
Hi @bip5 ,
> Edit: I've now transferred my data into the required format except for changing the labels from (0,1,2,4] to (0,1,2,3). Is this necessary with the tutorial pipeline? Which portion of the pipeline do I need to change to achieve this switch at data loading?

How many classes does the brats 2021 dataset have? You may need to modify the hyper-parameters of the network in: https://github.com/Project-MONAI/tutorials/blob/b429c883dccde3bfaa16c7c42fc2a35a8157fc3f/modules/dynunet_pipeline/create_network.py#L52

**Reply by bip5** [reply]
There's 4 classes, 1 negative and 3 positive. 

For an update, I changed the segmentation mask separately to have 0,1,2,3 instead of 0,1,2,4 and the training is working. I was hoping to be able to achieve this through the data loading pipeline. 

Ordinarily, I'd apply the transform inside __getitem__ for the dataset.   I'm not sure how I'd apply this change with this pipeline since it uses a CacheDataset and returns a dataloader instead of a dict 
 https://github.com/Project-MONAI/tutorials/blob/b429c883dccde3bfaa16c7c42fc2a35a8157fc3f/modules/dynunet_pipeline/create_dataset.py#L77

**Reply by yiheng-wang-nv** [reply]
Hi @bip5 , you may need to extend the dataset class and write a customized one, for example: https://github.com/Project-MONAI/tutorials/blob/b429c883dccde3bfaa16c7c42fc2a35a8157fc3f/kaggle/RANZCR/4th_place_solution/data/seg_data.py#L14
